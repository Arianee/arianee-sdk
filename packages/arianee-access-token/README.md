# @arianee/ArianeeAccessToken

The ArianeeAccessToken class is a module that allows you to create and manage Arianee Access Tokens (AAT). This class requires the Core class from the @arianee/core module

## Usage

You need to instanciate the class with a core instance and a wallet address.

```typescript
const core = core.fromPrivateKey('0x...');
const arianeeAccessToken = new ArianeeAccessToken(core);
```

Then you can use the following methods:

### `getValidWalletAccessToken(payloadOverride: PayloadOverride = {}, params?: { timeBeforeExp?: number; prefix?: string; }): Promise<string>`

This method generates a wallet scoped Arianee Access Token (AAT) and stores it in memory. On subsequent calls, if the stored AAT is still valid, it will return it. Otherwise if it has expired or the expiration is in less than `timeBeforeExp` seconds, it will regenerate a new one and return it.

You can use the `prefix` parameter to add a string before the arianee access token payload in the message to be signed.

### `createWalletAccessToken(payloadOverride: PayloadOverride = {}, prefix?: string): Promise<string>`

This method generates an Arianee Access Token (AAT) for the wallet scope. It returns a `Promise` that resolves to the AAT as a `string`.
It takes two optional parameters, a `payloadOverride` parameter to override the default payload and a `prefix` parameter to add a string before the arianee access token payload in the message to be signed.

### `createCertificateArianeeAccessToken(certificateId: number, network: string): Promise<string>`

This method generates an Arianee Access Token (AAT) for the certificate scope. It takes two parameters: `certificateId`, which is the ID of the certificate, and `network`, which is the name of the Arianee network. It returns a `Promise` that resolves to the AAT as a `string`.

### `createActionArianeeAccessTokenLink(url: string, certificateId: number, network: string): Promise<string>`

This method creates a link with an Arianee Access Token (AAT) attached to it. It takes three parameters: `url`, which is the URL to attach the AAT to, `certificateId`, which is the ID of the certificate, and `network`, which is the name of the Arianee network. It returns a `Promise` that resolves to the link with the AAT as a `string`.

<br>

If you only need to decode existing arianee access token, you don't need to instanciate the class.

You can use the following static methods. These methods will automatically detect if the arianee access token is prefixed and handle it. In order for this to work seamlessly, the arianee access tokens must be signed with one of these two signature algorithms (alg prop in header):Â `secp256k1` or `ETH`.

### `static isArianeeAccessTokenValid(arianeeAccessToken: string): boolean`

This static method checks if an Arianee Access Token (AAT) is valid. It takes an `arianeeAccessToken` parameter as a `string` and returns a `boolean` indicating whether the AAT is valid or not.

### `static decodeJwt(arianeeAccessToken: string): {header: JwtHeaderInterface, payload: ArianeeAccessTokenPayload, signature: string}`

This static method decodes an Arianee Access Token (AAT). It takes an `arianeeAccessToken` parameter as a `string` and returns an object with the decoded AAT, containing the header, payload, and signature.

## Storage

You can pass a `storage` object (that implements the [`Web Storage API`](https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API)) to the constructor under the `params.storage` key. \
This storage will be used for caching the generated arianee access token. Make sure that you do not pass an unsafe / public storage as it may expose the generated arianee access token.
