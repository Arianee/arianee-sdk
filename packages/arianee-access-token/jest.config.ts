/* eslint-disable */
export default {
  displayName: 'arianee-access-token',
  preset: '../../jest.preset.js',
  testEnvironment: 'jest-environment-node',
  transform: {
    '^.+\\.[tj]s$': ['ts-jest', { tsconfig: '<rootDir>/tsconfig.spec.json' }],
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: '../../coverage/packages/arianee-access-token',
};
