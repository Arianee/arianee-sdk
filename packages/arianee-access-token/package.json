{
  "name": "@arianee/arianee-access-token",
  "version": "1.89.0"
}
