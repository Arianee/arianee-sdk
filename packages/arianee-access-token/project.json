{
  "name": "arianee-access-token",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/arianee-access-token/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/arianee-access-token",
        "main": "packages/arianee-access-token/src/index.ts",
        "tsConfig": "packages/arianee-access-token/tsconfig.lib.json",
        "assets": ["packages/arianee-access-token/*.md"]
      }
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs arianee-access-token {args.ver} {args.tag}"
      },
      "dependsOn": ["build"]
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/arianee-access-token/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "packages/arianee-access-token/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  },
  "tags": []
}
