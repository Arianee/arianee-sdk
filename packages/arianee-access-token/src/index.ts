export * from './lib/arianee-access-token';

export * from './lib/types/JwtHeaderInterface';
export * from './lib/types/ArianeeAccessTokenPayload';
export * from './lib/types/SmartAssetSharingTokenPayload';
