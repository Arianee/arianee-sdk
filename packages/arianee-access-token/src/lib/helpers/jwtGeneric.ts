import { ArianeeAccessTokenPayload } from '../types/ArianeeAccessTokenPayload';
import { JwtHeaderInterface } from '../types/JwtHeaderInterface';

export class JWTGeneric {
  private static readonly JWT_HEADER_ETH = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJFVEgifQ';

  private static readonly JWT_HEADER_secp256k1 =
    'eyJ0eXAiOiJKV1QiLCJhbGciOiJzZWNwMjU2azEifQ';

  private header = { typ: 'JWT', alg: 'secp256k1' };
  private payload!: ArianeeAccessTokenPayload;
  private encodedToken!: string;

  constructor(
    private params: {
      signer?: (data: string) => Promise<string> | string;
      recover?: (message: string, signature: string) => string;
    }
  ) {}

  /**
   * Set payload to be signed
   * @param payload
   */
  public setPayload = async (
    payload: ArianeeAccessTokenPayload
  ): Promise<JWTGeneric> => {
    this.payload = payload;
    return this;
  };

  /**
   * Set payload to be signed
   * @param payload
   */
  public setHeader = async (payload: JwtHeaderInterface) => {
    this.header = payload;
    return this;
  };

  /**
   * Set token to be verified or decoded
   * @param encodedToken
   */
  public setToken = (encodedToken: string) => {
    this.encodedToken = encodedToken;
    return this;
  };

  private static base64Stringified(
    data: ArianeeAccessTokenPayload | JwtHeaderInterface
  ): string {
    const stringified = JSON.stringify(data);
    // https://datatracker.ietf.org/doc/html/rfc7515#appendix-C
    return Buffer.from(stringified).toString('base64');
    /* TO uncomment when sdk is deployed everywhere because we have a compatibility issue with the token
    .replace(/[=]/g, '')
    .replace(/\+/g, '-')
    .replace(/\//g, '_');
    */
  }

  private static fromBase64JSONParse(data: string) {
    let s = data;
    s = s.replace(/-/g, '+');
    s = s.replace(/_/g, '/');
    switch (s.length % 4) {
      case 0:
        break;
      case 2:
        s += '==';
        break;
      case 3:
        s += '=';
        break;
      default:
        throw new Error('Illegal base64url string!');
    }
    const buffer = Buffer.from(s, 'base64');
    const string = buffer.toString('utf8');
    return JSON.parse(string);
  }

  public async sign(prefix?: string) {
    const header = JWTGeneric.base64Stringified(this.header);
    const payload = JWTGeneric.base64Stringified(this.payload);
    const signature = await this.signature(prefix ?? '');
    return `${prefix ?? ''}${header}.${payload}.${signature}`;
  }

  /**
   * Verify if signature was signed by pubKey and return true/false
   * @param pubKey
   */
  public verify = (pubKey: string, timeBeforeExp?: number): boolean => {
    if (!this.params.recover) {
      throw new Error('You should provide a decoder to verify your token');
    }

    const [header, payload, signature] = this.encodedToken.split('.');
    const signedMessage = header + '.' + payload;
    const decode = this.params.recover(signedMessage, signature);

    const arePropertyValid = this.arePropertiesValid(timeBeforeExp);

    if (!arePropertyValid) {
      return false;
    }
    return pubKey.toLowerCase() === decode.toLowerCase();
  };

  /**
   * Check if expiration is before now + timeBeforeExpInSec.
   * If timeBeforeExpInSec === -1 we skip the expiration check
   * ex: does my token expires in the next 10 minutes
   * @param timeBeforeExpInSec
   * @returns
   */
  public isExpValid = (timeBeforeExpInSec = 0): boolean => {
    if (timeBeforeExpInSec === -1) {
      return true;
    }
    const decoded = this.decode();
    const now = new Date().getTime();
    const isExpInMilliseconds = decoded.payload.exp.toString().length === 13;
    const expInMilliseconds = isExpInMilliseconds
      ? decoded.payload.exp
      : decoded.payload.exp * 1000;

    return now + timeBeforeExpInSec * 1000 < expInMilliseconds;
  };

  public arePropertiesValid = (timeBeforeExpInSec = 0): boolean => {
    const { payload } = this.decode();

    const isExpired = !this.isExpValid(timeBeforeExpInSec);
    if (isExpired) {
      return false;
    }

    if (payload.nbf) {
      const nbfInMS = payload.nbf * 1000;
      const isBefore = new Date(nbfInMS).getTime() > Date.now();
      if (isBefore) {
        return false;
      }
    }

    return true;
  };

  public decode = (): {
    prefix: string;
    header: JwtHeaderInterface;
    payload: ArianeeAccessTokenPayload;
    signature: string;
  } => {
    let headerType = this.encodedToken.includes(JWTGeneric.JWT_HEADER_ETH)
      ? JWTGeneric.JWT_HEADER_ETH
      : this.encodedToken.includes(JWTGeneric.JWT_HEADER_secp256k1)
      ? JWTGeneric.JWT_HEADER_secp256k1
      : null;

    // AAT used to not match RFC 7515. But we need to be retro-compatible until fix is used everywhere
    if (this.encodedToken.includes('==')) {
      headerType = headerType + '==';
    }

    let prefix;
    let remainder;
    let header, payload, signature;
    if (headerType) {
      [prefix, remainder] = this.encodedToken.split(`${headerType}.`);
      [header, payload, signature] = [headerType, ...remainder.split('.')];
    } else {
      remainder = this.encodedToken;
      [header, payload, signature] = remainder.split('.');
    }

    return {
      prefix: prefix ?? '',
      header: JWTGeneric.fromBase64JSONParse(
        header
      ) as unknown as JwtHeaderInterface,
      payload: JWTGeneric.fromBase64JSONParse(
        payload
      ) as unknown as ArianeeAccessTokenPayload,
      signature: signature,
    };
  };

  private signature(prefix: string) {
    if (!this.params.signer) {
      throw new Error('You must provide a signer');
    }

    return this.params.signer(
      prefix +
        JWTGeneric.base64Stringified(this.header) +
        '.' +
        JWTGeneric.base64Stringified(this.payload)
    );
  }
}
