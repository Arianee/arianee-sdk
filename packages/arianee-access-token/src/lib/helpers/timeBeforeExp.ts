import { ArianeeAccessToken } from '../arianee-access-token';

export function isExpInLessThan(
  aat: string,
  timeBeforeExpInSec: number
): boolean {
  const decoded = ArianeeAccessToken.decodeJwt(aat);
  const now = new Date().getTime();
  return decoded.payload.exp - now < timeBeforeExpInSec * 1000;
}
