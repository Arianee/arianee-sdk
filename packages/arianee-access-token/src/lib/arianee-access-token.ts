import { Core } from '@arianee/core';
import { ArianeeAccessTokenPayload } from './types/arianeeAccessTokenPayload';
import { JWTGeneric } from './helpers/jwtGeneric';
import { ethers } from 'ethers';
import { JwtHeaderInterface } from './types/JwtHeaderInterface';
import { isExpInLessThan } from './helpers/timeBeforeExp';

interface PayloadOverride {
  exp?: number;
  iat?: number;
}

export class ArianeeAccessToken {
  private lastAAT!: string;

  constructor(private core: Core) {}

  public async getValidWalletAccessToken(
    payloadOverride: PayloadOverride = {},
    timeBeforeExp = 10
  ): Promise<string> {
    if (!this.lastAAT || isExpInLessThan(this.lastAAT, timeBeforeExp)) {
      this.lastAAT = await this.createWalletAccessToken(payloadOverride);
    }
    return this.lastAAT;
  }

  public createWalletAccessToken(
    payloadOverride: PayloadOverride = {}
  ): Promise<string> {
    return this.generateAAT(payloadOverride);
  }

  public createCertificateArianeeAccessToken(
    certificateId: number,
    network: string,
    payloadOverride: PayloadOverride = {}
  ): Promise<string> {
    return this.generateAAT({
      subId: certificateId,
      sub: 'certificate',
      network: network,
      ...payloadOverride,
    });
  }

  public async createActionArianeeAccessTokenLink(
    url: string,
    certificateId: number,
    network: string
  ): Promise<string> {
    const urlObject = new URL(url);
    const arianeeAccessToken = await this.createCertificateArianeeAccessToken(
      certificateId,
      network
    );
    urlObject.searchParams.append('arianeeAccessToken', arianeeAccessToken);
    return urlObject.toString();
  }

  static isArianeeAccessTokenValid(
    arianeeAccessToken: string,
    ignoreExpiration = false
  ): boolean {
    const recover = (message: string, signature: string): string =>
      ethers.verifyMessage(message, signature);
    const jwtGenerator = new JWTGeneric({ recover });
    const jwt = jwtGenerator.setToken(arianeeAccessToken);
    const iss = jwt.decode().payload.iss;

    return jwt.verify(iss, ignoreExpiration);
  }

  static decodeJwt(
    arianeeAccessToken: string,
    ignoreExpiration = false
  ): {
    header: JwtHeaderInterface;
    payload: ArianeeAccessTokenPayload;
    signature: string;
  } {
    const isAatValid = ArianeeAccessToken.isArianeeAccessTokenValid(
      arianeeAccessToken,
      ignoreExpiration
    );
    if (!isAatValid) {
      throw new Error('ArianeeAccessToken is not valid');
    }
    const jwtGenerator = new JWTGeneric({});
    const jwt = jwtGenerator.setToken(arianeeAccessToken);
    return jwt.decode();
  }

  private async generateAAT(
    payload: Partial<ArianeeAccessTokenPayload> = {}
  ): Promise<string> {
    const signer = async (data: string) => {
      return (await this.core.signMessage(data)).signature;
    };
    const jwtGenerator = new JWTGeneric({ signer });
    const now = Date.now();
    const basicPayload: ArianeeAccessTokenPayload = {
      iss: this.core.getAddress(),
      sub: 'wallet',
      exp: now + 5 * 60 * 1000,
      iat: now,
      ...payload,
    };

    const jwt = await jwtGenerator.setPayload(basicPayload);
    return jwt.sign();
  }
}
