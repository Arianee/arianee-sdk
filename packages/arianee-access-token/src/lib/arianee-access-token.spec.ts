/* eslint-disable @typescript-eslint/no-explicit-any */
import { Core } from '@arianee/core';
import { MemoryStorage } from '@arianee/utils';

import { ArianeeAccessToken } from './arianee-access-token';
import * as timeBeforeExpModule from './helpers/timeBeforeExp';

describe('arianeeAccessToken', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const pk =
    '0xaab1af774ae54a2efaff0d3d93308c00fa7b639304da92847cfc986af4b87eb5';
  const core = Core.fromPrivateKey(pk);

  describe('constructor', () => {
    it('should use the MemoryStorage if no storage passed', () => {
      const arianeeAccessToken = new ArianeeAccessToken(core);
      expect(arianeeAccessToken['storage']).toBeInstanceOf(MemoryStorage);
    });

    it('should use the passed storage', () => {
      const storage = new MemoryStorage();

      const arianeeAccessToken = new ArianeeAccessToken(core, {
        storage,
      });

      expect(arianeeAccessToken['storage']).toBe(storage);
    });

    it('should set the initial walletAccessToken if passed', () => {
      const storage = {
        setItem: jest.fn(),
      } as unknown as Storage;

      const arianeeAccessToken = new ArianeeAccessToken(core, {
        storage,
        initialValues: {
          walletAccessToken: 'mockToken',
        },
      });

      expect(storage.setItem).toHaveBeenCalledWith(
        'arianee__lastAAT',
        'mockToken'
      );
      expect(arianeeAccessToken['storage']).toBe(storage);
    });
  });

  it('should generate a valid aat', async () => {
    const aatGenerator = new ArianeeAccessToken(core);
    const aat = await aatGenerator['generateAAT']({ iat: 0, exp: 0 });
    expect(aat).toEqual(
      'eyJ0eXAiOiJKV1QiLCJhbGciOiJzZWNwMjU2azEifQ==.eyJpc3MiOiIweDc2OTFlMDcwNUMyRThlRDc5MzY2QjA2N2Q1ZkNBRmE4OUFBMDdGODgiLCJzdWIiOiJ3YWxsZXQiLCJleHAiOjAsImlhdCI6MH0=.0x6ec31d56cb6ec7ff9cf9c04f16c4bdc75b9b1d1b35d1940f20b3e905e162600b67a1970f537b9ae8c26c1f05a93913d13d599923880f71ed1198b331593bd8f81b'
    );
  });

  it('should verify aat validity', async () => {
    // this aat has a long exp date
    const aat =
      'eyJ0eXAiOiJKV1QiLCJhbGciOiJFVEgifQ==.eyJpc3MiOiIweDc2OTFlMDcwNUMyRThlRDc5MzY2QjA2N2Q1ZkNBRmE4OUFBMDdGODgiLCJzY29wZSI6ImFsbCIsImV4cCI6NjAxNjgyMjQwNzQxOTg4LCJpYXQiOjE2ODIyNDA3NDE5ODgsImNoYWluIjoidGVzdG5ldCIsInN1YiI6IndhbGxldCJ9.0x7bfc70360ed7809d2b0c9e348930c10917ebcabfee5cfa0dc608719e8eaae812492c497a388fce5ff61ef5fb9a1c13ac4ef33075249e8c63adc95e05603188b21b';
    const aatValidity = ArianeeAccessToken.isArianeeAccessTokenValid(aat);
    expect(aatValidity).toEqual(true);
  });

  it('should verify a prefixed aat validity', async () => {
    // this aat has a long exp date
    const prefixedAat =
      'Please sign this arianee access token\neyJ0eXAiOiJKV1QiLCJhbGciOiJzZWNwMjU2azEifQ==.eyJpc3MiOiIweDc2OTFlMDcwNUMyRThlRDc5MzY2QjA2N2Q1ZkNBRmE4OUFBMDdGODgiLCJzdWIiOiJ3YWxsZXQiLCJleHAiOjYwMTY4MjI0MDc0MTk4OCwiaWF0IjoxNjg2MzAzNzg1NzE0fQ==.0x034fdc0cef784787bbc085adc73c1f9adb59e6045bfcb3aab7caeeb5e939e26d266e6eb84956df6c0517b4c61d623e6e3ab5c84d13fb3d9c7a27dba811168ac81c';
    const isValid = ArianeeAccessToken.isArianeeAccessTokenValid(prefixedAat);
    expect(isValid).toEqual(true);
  });

  it('should decode a prefixed aat', async () => {
    // this aat has a long exp date
    const prefixedAat =
      'Please sign this arianee access token\neyJ0eXAiOiJKV1QiLCJhbGciOiJzZWNwMjU2azEifQ==.eyJpc3MiOiIweDc2OTFlMDcwNUMyRThlRDc5MzY2QjA2N2Q1ZkNBRmE4OUFBMDdGODgiLCJzdWIiOiJ3YWxsZXQiLCJleHAiOjIwMDE2NjM5NjkwOTUsImlhdCI6MTY4NjMwMzk2OTA5NX0=.0x3d7a5471eb857cf5b0f0893e4d78ec7dbcfc0b3670a4a366b049abd7e4517b3a2ebc04b6b4661e75762c3d324ec274718fb1ca49203d0fb01e4d71a4c9a5c2e51c';

    const decoded = ArianeeAccessToken.decodeJwt(prefixedAat);

    expect(decoded.payload).toMatchObject({
      iss: '0x7691e0705C2E8eD79366B067d5fCAFa89AA07F88',
      sub: 'wallet',
      exp: 2001663969095,
      iat: 1686303969095,
    });
  });

  it('should create a validate an aat', async () => {
    const aatGenerator = new ArianeeAccessToken(core);
    const aat = await aatGenerator['generateAAT']();
    const aatValidity = ArianeeAccessToken.isArianeeAccessTokenValid(aat);
    expect(aatValidity).toEqual(true);
  });

  it('should create a validate an aat', async () => {
    const aatGenerator = new ArianeeAccessToken(core);
    const aat = await aatGenerator['generateAAT']();
    const decoded = ArianeeAccessToken.decodeJwt(aat);
    expect(decoded.payload.iss).toEqual(core.getAddress());
  });

  describe('getValidWalletAccessToken', () => {
    it('returns the existing token if it is valid', async () => {
      const aatGenerator = new ArianeeAccessToken(core);
      const spy = jest.spyOn(aatGenerator, 'createWalletAccessToken');

      const aat1 = await aatGenerator.getValidWalletAccessToken();
      const aat2 = await aatGenerator.getValidWalletAccessToken();
      expect(aat1).toBe(aat2);
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('returns the passed initial token if it is valid', async () => {
      const initialAAT =
        'eyJ0eXAiOiJKV1QiLCJhbGciOiJzZWNwMjU2azEifQ==.eyJpc3MiOiIweDc2OTFlMDcwNUMyRThlRDc5MzY2QjA2N2Q1ZkNBRmE4OUFBMDdGODgiLCJzdWIiOiJ3YWxsZXQiLCJleHAiOjE4NzQ3NDQ3MDY2NjEsImlhdCI6MTcwMTk0NDcwNjY2MX0=.0xc325b872a4ea65b89ba8846685920b90155fcb8c9d87c06dd4fcdbc708ad4f1a4dc67029b946f44284d0442eee03b08ee9bba1de5bfd6769fd2bbd86363456b41c';

      const aatGenerator = new ArianeeAccessToken(core, {
        initialValues: {
          walletAccessToken: initialAAT,
        },
      });

      const spy = jest.spyOn(aatGenerator, 'createWalletAccessToken');

      const aat1 = await aatGenerator.getValidWalletAccessToken();
      const aat2 = await aatGenerator.getValidWalletAccessToken();
      expect(aat1).toBe(aat2);
      expect(aat1).toBe(initialAAT);
      expect(spy).not.toHaveBeenCalled();
    });

    it('creates a new token if the existing token is expired', async () => {
      const aatGenerator = new ArianeeAccessToken(core);
      const spy = jest.spyOn(aatGenerator, 'createWalletAccessToken');
      const isExpInLessThanSpy = jest.spyOn(
        timeBeforeExpModule,
        'isExpInLessThan'
      );

      const aat1 = await aatGenerator.getValidWalletAccessToken({ exp: 0 });
      const aat2 = await aatGenerator.getValidWalletAccessToken();

      expect(aat1).not.toBe(aat2);
      expect(spy).toHaveBeenCalledTimes(2);
      expect(isExpInLessThanSpy).toHaveBeenCalledWith(aat1, 10);
    });

    it('creates a new token if there is no existing token', async () => {
      const aatGenerator = new ArianeeAccessToken(core);
      const spy = jest.spyOn(aatGenerator, 'createWalletAccessToken');

      const aat1 = await aatGenerator.getValidWalletAccessToken();
      expect(aat1).toBeDefined();
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('creates a new token with the passed prefix', async () => {
      const aat = new ArianeeAccessToken(core);
      const createWalletAccessTokenSpy = jest.spyOn(
        aat,
        'createWalletAccessToken'
      );

      const prefix = 'Please sign this arianee access token\n';

      const generatedAat = await aat.getValidWalletAccessToken({}, { prefix });

      expect(generatedAat).toMatch(new RegExp(prefix + '(a-zA-Zd.)*', 'gi'));
      expect(createWalletAccessTokenSpy).toHaveBeenCalledTimes(1);
      expect(createWalletAccessTokenSpy).toHaveBeenCalledWith({}, prefix);
    });
  });

  describe('createWalletAccessToken', () => {
    it('should call and return generateAAT with correct params', async () => {
      const aat = new ArianeeAccessToken(core);

      const generateAATSpy = jest.spyOn(aat as any, 'generateAAT');
      const generatedAat = await aat.createWalletAccessToken({}, 'prefix');

      expect(generateAATSpy).toHaveBeenCalledWith({}, 'prefix');
      expect(generatedAat).toMatch(/prefix(a-zA-Zd.)*/gi);
    });
  });
});
