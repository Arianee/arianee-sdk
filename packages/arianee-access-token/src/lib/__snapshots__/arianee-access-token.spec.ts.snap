// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arianeeAccessToken decodeJWTWithoutValidation should decode an non aat jwt 1`] = `
{
  "header": {
    "alg": "HS256",
    "typ": "JWT",
  },
  "payload": {
    "iat": 1516239022,
    "name": "John Doe",
    "sub": "1234567890",
  },
  "prefix": "",
  "signature": "SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c",
}
`;

exports[`arianeeAccessToken should generate a valid aat 1`] = `"eyJ0eXAiOiJKV1QiLCJhbGciOiJzZWNwMjU2azEifQ==.eyJpc3MiOiIweDc2OTFlMDcwNUMyRThlRDc5MzY2QjA2N2Q1ZkNBRmE4OUFBMDdGODgiLCJzdWIiOiJ3YWxsZXQiLCJleHAiOjAsImlhdCI6MH0=.0x6ec31d56cb6ec7ff9cf9c04f16c4bdc75b9b1d1b35d1940f20b3e905e162600b67a1970f537b9ae8c26c1f05a93913d13d599923880f71ed1198b331593bd8f81b"`;
